### PR TITLE
Enhancement/ Add new overlay for opponent view when seven double jack…

### PR DESCRIPTION
**Pull request by TommySchmid**
… edge case occurs

## Please check the following

- [x] Do the tests still pass? (see [Run the Tests](https://github.com/cuttle-cards/cuttle#run-the-tests))
- [x] Is the code formatted properly? (see [Linting (Formatting)](https://github.com/cuttle-cards/cuttle#linting-formatting))
- For New Features:
  - [x] Have tests been added to cover any new features or fixes?
  - [x] Has the documentation been updated accordingly?

### CODE REVIEW

File: tests/e2e/specs/in-game/sevens.spec.js [hunk 1]
```diff
@@ -797,7 +797,7 @@ describe('Opponent playing SEVENS', () => {
     cy.get('[data-cy=cannot-counter-resolve]').should('be.visible').click();
     cy.log('Player resolves (could not counter');
 
-    // Waiting for opponent
+    // // Waiting for opponent
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/sevens.spec.js:800]:
```suggestion
    // Waiting for opponent
```

File: tests/e2e/specs/in-game/sevens.spec.js [hunk 2]
```diff
@@ -816,10 +816,10 @@ describe('Opponent playing SEVENS', () => {
       .click({ force: true }); // can't play to scrap
     cy.get('#player-field').should('not.have.class', 'valid-move').click({ force: true }); // can't play to field
 
-    // Opponent plays four of clubs for points
+    // // Opponent plays four of clubs for points
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/sevens.spec.js:819]:
```suggestion
    // Opponent plays four of clubs for points
```

File: tests/e2e/specs/in-game/sevens.spec.js [hunk 3]
```diff
@@ -816,10 +816,10 @@ describe('Opponent playing SEVENS', () => {
...
     cy.playPointsFromSevenOpponent(Card.FOUR_OF_CLUBS);
 
-    // No longer waiting for opponent
+    // // No longer waiting for opponent
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/sevens.spec.js:822]:
```suggestion
    // No longer waiting for opponent
```

File: client/js/components/GameView/GameOverlays.vue [hunk 4]
```diff
@@ -104,6 +113,26 @@ export default {
       }
       return mayCounter;
     },
+    showWaitingForOpponentToDiscardJackFromDeck() {
+      return (
+        this.resolvingSeven &&
```
Comment (itsalaidbacklife) [inline client/js/components/GameView/GameOverlays.vue:118]: So did you say that you tried replacing `resolvingSeven` with the more specific `waitingForOpponentToPlayFromDeck` and it didn't work?
```suggestion
        this.waitingForOpponentToPlayFromDeck &&
```

File: client/js/components/GameView/GameOverlays.vue [hunk 5]
```diff
@@ -104,6 +113,26 @@ export default {
...
+        this.topTwoCardsCantBePlayed &&
+        (this.opponentPointTotal === 0 || this.opponentQueenCount > 0) &&
+        this.isPlayersTurn === false
```
Comment (itsalaidbacklife) [inline client/js/components/GameView/GameOverlays.vue:121]: Using `waitingForOpponentToPlayFromDeck` above should alleviate the need for the turn check b/c we only set `waitingForOpponentToPlayFromDeck` when it is the opponent's turn
```suggestion
        (this.opponentPointTotal === 0 || this.opponentQueenCount > 0)
```

File: client/js/components/GameView/GameDialogs.vue [hunk 6]
```diff
@@ -121,7 +122,8 @@ export default {
         this.resolvingSeven &&
```
Comment (itsalaidbacklife) [inline client/js/components/GameView/GameDialogs.vue:122]: We can use `playingFromDeck` instead of `resolvingSeven` because `playingFromDeck` gets set when you're resolving a seven _and_ it's your turn
```suggestion
        this.playingFromDeck &&
```

File: client/js/components/GameView/GameDialogs.vue [hunk 7]
```diff
@@ -121,7 +122,8 @@ export default {
...
         this.topCard.rank === 11 &&
         (!this.secondCard || this.secondCard.rank === 11) &&
-        (this.opponentPointTotal === 0 || this.opponentQueenCount > 0)
+        (this.opponentPointTotal === 0 || this.opponentQueenCount > 0) &&
+        this.isPlayersTurn === true
```
Comment (itsalaidbacklife) [inline client/js/components/GameView/GameDialogs.vue:126]: Then we could remove the turn check here
```suggestion
        (this.opponentPointTotal === 0 || this.opponentQueenCount > 0)
```

File: tests/e2e/specs/in-game/sevens.spec.js (hunk 1 above)
Comment (TommySchmid) [reply tests/e2e/specs/in-game/sevens.spec.js:800]: Arrrggghh you beat me to this haha

File: client/js/components/GameView/GameOverlays.vue (hunk 4 above)
Comment (TommySchmid) [reply client/js/components/GameView/GameOverlays.vue:118]: Removing the resolvingSeven check completely displays the overlay whenever two jacks are on top of the deck with no points, without the seven being played.

Comment (itsalaidbacklife) [reply client/js/components/GameView/GameOverlays.vue:118]: Are you talking about removing it completely, or replacing it with `waitingForOpponentToPlayFromDeck`?

Comment (itsalaidbacklife) [reply client/js/components/GameView/GameOverlays.vue:118]: My suggestion isn't to remove it, it's to check a different variable

File: tests/e2e/specs/in-game/sevens.spec.js (hunk 1 above)
Comment (itsalaidbacklife) [reply tests/e2e/specs/in-game/sevens.spec.js:800]: ;)

File: client/js/components/GameView/GameDialogs.vue (hunk 6 above)
Comment (TommySchmid) [reply client/js/components/GameView/GameDialogs.vue:122]: This change does the trick for the opponent view, but causes tests for playing a seven into a double jack scenario to fail.

File: tests/e2e/support/commands.js [hunk 8]
```diff
@@ -712,7 +712,11 @@ Cypress.Commands.add('playJackFromSevenOpponent', (card, target) => {
     .then((game) => {
       const player = game.players[game.myPNum];
       let foundCard;
-      const foundTarget = player.points.find((pointCard) => cardsMatch(target, pointCard));
+
+      const discarding = target === -1;
+      const foundTarget = discarding
+        ? -1
+        : player.points.find((pointCard) => cardsMatch(target, pointCard));
```
Comment (itsalaidbacklife) [inline tests/e2e/support/commands.js:719]: Okay looking at it more closely, maybe instead of a ternary it'd make sense to use a regular conditional like this
```suggestion
      let targetId = -1;
      if (!discarding) {
        const foundTarget = player.points.find((pointCard) => cardsMatch(target, pointCard));
        if (!foundTarget) {
          throw new Error(
            `Error playing opponents jack: could not find ${target.rank} of ${target.suit} in player points`
          );
        }
        targetId = foundTarget.id
      }
```

Comment (itsalaidbacklife) [reply tests/e2e/support/commands.js:719]: Do you think this is any more clear?. Then we could remove some of the stuff below

File: tests/e2e/support/commands.js [hunk 9]
```diff
@@ -735,7 +739,14 @@ Cypress.Commands.add('playJackFromSevenOpponent', (card, target) => {
...
       const cardId = foundCard.id;
-      const targetId = foundTarget.id;
+      let targetId;
+
+      if (target !== -1) {
+        targetId = foundTarget.id;
+      } else {
+        targetId = -1;
+      }
+
```
Comment (itsalaidbacklife) [inline tests/e2e/support/commands.js:749]: If we do the above we'd remove this (b/c it's moved above)

File: tests/e2e/support/commands.js [hunk 10]
```diff
@@ -735,7 +739,14 @@ Cypress.Commands.add('playJackFromSevenOpponent', (card, target) => {
...
 
       const cardId = foundCard.id;
-      const targetId = foundTarget.id;
+      let targetId;
+
+      if (target !== -1) {
+        targetId = foundTarget.id;
+      } else {
+        targetId = -1;
+      }
```
Comment (itsalaidbacklife) [inline tests/e2e/support/commands.js:748]: Same with this
```suggestion
      let targetId;

      if (target !== -1) {
        targetId = foundTarget.id;
      } else {
        targetId = -1;
      }
```

File: tests/e2e/support/commands.js (hunk 8 above)
Comment (itsalaidbacklife) [reply tests/e2e/support/commands.js:719]: I'll leave it to your judgement. What do you think is clearest?

File: client/js/components/GameView/GameDialogs.vue (hunk 6 above)
Comment (itsalaidbacklife) [reply client/js/components/GameView/GameDialogs.vue:122]: That surprises me, huh. I wonder why

File: tests/e2e/support/commands.js (hunk 8 above)
Comment (TommySchmid) [reply tests/e2e/support/commands.js:719]: I think the ternary operator is sufficient here for the time being. Since you said you wanted to eventually rethink the -1 usage altogether, we could also revisit this then.

File: client/js/components/GameView/GameDialogs.vue (hunk 6 above)
Comment (TommySchmid) [reply client/js/components/GameView/GameDialogs.vue:122]: ^^ Per our conversation, I added `playingFromDeck: ({ game }) => game.playingFromDeck` to a new mapState expression in GameDialogs. Works as intended

File: client/js/components/GameView/GameOverlays.vue [hunk 11]
```diff
@@ -104,6 +113,25 @@ export default {
       }
       return mayCounter;
     },
+    showWaitingForOpponentToDiscardJackFromDeck() {
+      return (
+        this.waitingForOpponentToPlayFromDeck &&
+        this.topTwoCardsCantBePlayed &&
+        (this.opponentPointTotal === 0 || this.opponentQueenCount > 0)
```
Comment (itsalaidbacklife) [inline client/js/components/GameView/GameOverlays.vue:120]: let's fold this into `topTwoCardsCantBePlayed` since that'd be more semantically accurate. You can't play the top two cards if they are both jacks _and_ jacks can't be played because of a queen or no points to target

File: client/js/components/GameView/GameOverlays.vue [hunk 12]
```diff
@@ -104,6 +113,25 @@ export default {
...
+    },
+    topTwoCardsCantBePlayed() {
+      return this.topCard.rank === 11 && (!this.secondCard || this.secondCard.rank === 11);
+    },
+    topCard() {
+      return this.$store.state.game.topCard;
+    },
+    secondCard() {
+      return this.$store.state.game.secondCard;
+    },
```
Comment (itsalaidbacklife) [inline client/js/components/GameView/GameOverlays.vue:134]: These two can be done with `mapState()` next to the `waitingForOpponent...` stuff

**Review by itsalaidbacklife — APPROVED** [review]
Great work. Requesting two last changes then we're good to merge

File: client/js/components/GameView/GameOverlays.vue [hunk 13]
```diff
@@ -5,7 +5,7 @@
       v-model="waitingForOpponentToCounter"
       opacity=".6"
     >
-      <h1>{{ waitingForOpponetToCounterMessage }}</h1>
+      <h1>{{ showWaitingForOpponetToCounterMessage }}</h1>
```
Comment (TommySchmid) [inline client/js/components/GameView/GameOverlays.vue:8]: Added the `show` here to help distinguish between computed values and values mapped from the state. I ran into some confusion using `waitingForOpponentToPlayFromDeck` when trying to make all computed props uniform in name. Since it's already a mapped value, things broke.

**Comment by itsalaidbacklife** [comment]
Resolves #226

**Comment by itsalaidbacklife** [comment]
Resolves #196
